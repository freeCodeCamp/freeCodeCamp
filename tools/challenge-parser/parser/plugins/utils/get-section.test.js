const isArray = require('lodash/isArray');
const { root } = require('mdast-builder');
const find = require('unist-util-find');

const parseFixture = require('../../__fixtures__/parse-fixture');
const { getSection } = require('./get-section');

describe('getSection', () => {
  let simpleAst, extraHeadingAst;

  beforeAll(async () => {
    extraHeadingAst = await parseFixture('with-extra-heading.md');
    simpleAst = await parseFixture('simple.md');
  });

  it('should return an array', () => {
    expect.assertions(1);
    const actual = getSection(simpleAst, '--hints--');
    expect(isArray(actual)).toBe(true);
  });

  it('should return an empty array if the marker is not present', () => {
    expect.assertions(2);
    const actual = getSection(simpleAst, '--not-a-marker--');
    expect(isArray(actual)).toBe(true);
    expect(actual.length).toBe(0);
  });

  it('should include any headings without markers', () => {
    expect.assertions(1);
    const actual = getSection(extraHeadingAst, '--description--');
    expect(
      find(root(actual), {
        value: 'this should still be inside --description--'
      })
    ).not.toBeUndefined();
  });

  it('should include the rest of the AST if there is no end marker', () => {
    expect.assertions(2);
    const actual = getSection(extraHeadingAst, '--solutions--');
    expect(actual.length > 0).toBe(true);
    expect(
      find(root(actual), { value: 'body {\n  background: white;\n}' })
    ).not.toBeUndefined();
  });

  it('should match the hints snapshot', () => {
    const actual = getSection(simpleAst, '--hints--');
    expect(actual).toMatchSnapshot();
  });
  it('should match the instructions snapshot', () => {
    const actual = getSection(simpleAst, '--instructions--');
    expect(actual).toMatchSnapshot();
  });
});
