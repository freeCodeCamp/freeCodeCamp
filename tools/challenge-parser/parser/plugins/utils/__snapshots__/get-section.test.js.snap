// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSection should match the hints snapshot 1`] = `
[
  {
    "children": [
      {
        "position": {
          "end": {
            "column": 11,
            "line": 19,
            "offset": 142,
          },
          "start": {
            "column": 1,
            "line": 19,
            "offset": 132,
          },
        },
        "type": "text",
        "value": "First hint",
      },
    ],
    "position": {
      "end": {
        "column": 11,
        "line": 19,
        "offset": 142,
      },
      "start": {
        "column": 1,
        "line": 19,
        "offset": 132,
      },
    },
    "type": "paragraph",
  },
  {
    "lang": "js",
    "meta": null,
    "position": {
      "end": {
        "column": 4,
        "line": 23,
        "offset": 166,
      },
      "start": {
        "column": 1,
        "line": 21,
        "offset": 144,
      },
    },
    "type": "code",
    "value": "// test code",
  },
  {
    "children": [
      {
        "position": {
          "end": {
            "column": 18,
            "line": 25,
            "offset": 185,
          },
          "start": {
            "column": 1,
            "line": 25,
            "offset": 168,
          },
        },
        "type": "text",
        "value": "Second hint with ",
      },
      {
        "position": {
          "end": {
            "column": 24,
            "line": 25,
            "offset": 191,
          },
          "start": {
            "column": 18,
            "line": 25,
            "offset": 185,
          },
        },
        "type": "html",
        "value": "<code>",
      },
      {
        "position": {
          "end": {
            "column": 28,
            "line": 25,
            "offset": 195,
          },
          "start": {
            "column": 24,
            "line": 25,
            "offset": 191,
          },
        },
        "type": "text",
        "value": "code",
      },
      {
        "position": {
          "end": {
            "column": 35,
            "line": 25,
            "offset": 202,
          },
          "start": {
            "column": 28,
            "line": 25,
            "offset": 195,
          },
        },
        "type": "html",
        "value": "</code>",
      },
    ],
    "position": {
      "end": {
        "column": 35,
        "line": 25,
        "offset": 202,
      },
      "start": {
        "column": 1,
        "line": 25,
        "offset": 168,
      },
    },
    "type": "paragraph",
  },
  {
    "lang": "js",
    "meta": null,
    "position": {
      "end": {
        "column": 4,
        "line": 29,
        "offset": 231,
      },
      "start": {
        "column": 1,
        "line": 27,
        "offset": 204,
      },
    },
    "type": "code",
    "value": "// more test code",
  },
  {
    "children": [
      {
        "position": {
          "end": {
            "column": 7,
            "line": 31,
            "offset": 239,
          },
          "start": {
            "column": 1,
            "line": 31,
            "offset": 233,
          },
        },
        "type": "text",
        "value": "Third ",
      },
      {
        "children": [
          {
            "position": {
              "end": {
                "column": 12,
                "line": 31,
                "offset": 244,
              },
              "start": {
                "column": 8,
                "line": 31,
                "offset": 240,
              },
            },
            "type": "text",
            "value": "hint",
          },
        ],
        "position": {
          "end": {
            "column": 13,
            "line": 31,
            "offset": 245,
          },
          "start": {
            "column": 7,
            "line": 31,
            "offset": 239,
          },
        },
        "type": "emphasis",
      },
      {
        "position": {
          "end": {
            "column": 19,
            "line": 31,
            "offset": 251,
          },
          "start": {
            "column": 13,
            "line": 31,
            "offset": 245,
          },
        },
        "type": "text",
        "value": " with ",
      },
      {
        "position": {
          "end": {
            "column": 25,
            "line": 31,
            "offset": 257,
          },
          "start": {
            "column": 19,
            "line": 31,
            "offset": 251,
          },
        },
        "type": "html",
        "value": "<code>",
      },
      {
        "position": {
          "end": {
            "column": 29,
            "line": 31,
            "offset": 261,
          },
          "start": {
            "column": 25,
            "line": 31,
            "offset": 257,
          },
        },
        "type": "text",
        "value": "code",
      },
      {
        "position": {
          "end": {
            "column": 36,
            "line": 31,
            "offset": 268,
          },
          "start": {
            "column": 29,
            "line": 31,
            "offset": 261,
          },
        },
        "type": "html",
        "value": "</code>",
      },
      {
        "position": {
          "end": {
            "column": 41,
            "line": 31,
            "offset": 273,
          },
          "start": {
            "column": 36,
            "line": 31,
            "offset": 268,
          },
        },
        "type": "text",
        "value": " and ",
      },
      {
        "position": {
          "end": {
            "column": 54,
            "line": 31,
            "offset": 286,
          },
          "start": {
            "column": 41,
            "line": 31,
            "offset": 273,
          },
        },
        "type": "inlineCode",
        "value": "inline code",
      },
    ],
    "position": {
      "end": {
        "column": 54,
        "line": 31,
        "offset": 286,
      },
      "start": {
        "column": 1,
        "line": 31,
        "offset": 233,
      },
    },
    "type": "paragraph",
  },
  {
    "lang": "js",
    "meta": null,
    "position": {
      "end": {
        "column": 4,
        "line": 38,
        "offset": 353,
      },
      "start": {
        "column": 1,
        "line": 33,
        "offset": 288,
      },
    },
    "type": "code",
    "value": "// more test code
if(let x of xs) {
  console.log(x);
}",
  },
]
`;

exports[`getSection should match the instructions snapshot 1`] = `
[
  {
    "children": [
      {
        "position": {
          "end": {
            "column": 12,
            "line": 11,
            "offset": 89,
          },
          "start": {
            "column": 1,
            "line": 11,
            "offset": 78,
          },
        },
        "type": "text",
        "value": "Paragraph 0",
      },
    ],
    "position": {
      "end": {
        "column": 12,
        "line": 11,
        "offset": 89,
      },
      "start": {
        "column": 1,
        "line": 11,
        "offset": 78,
      },
    },
    "type": "paragraph",
  },
  {
    "lang": "html",
    "meta": null,
    "position": {
      "end": {
        "column": 4,
        "line": 15,
        "offset": 117,
      },
      "start": {
        "column": 1,
        "line": 13,
        "offset": 91,
      },
    },
    "type": "code",
    "value": "code example 0",
  },
]
`;
